src/main/scala/viper/silver/plugin/standard/smoke/SmokeDetectionPlugin.scala

```diff
@@ -0,0 +1,215 @@
+package viper.silver.plugin.standard.smoke
+
+import fastparse.P
+import viper.silver.ast._
+import viper.silver.ast.utility.ViperStrategy
+import viper.silver.cfg.silver.SilverCfg
+import viper.silver.cfg.silver.SilverCfg.SilverBlock
+import viper.silver.cfg.utility.IdInfo
+import viper.silver.parser.FastParser
+import viper.silver.plugin.standard.refute.Refute
+import viper.silver.plugin.{ParserPluginTemplate, SilverPlugin}
+
+import scala.annotation.unused
+import scala.collection.mutable.ListBuffer
+
+class SmokeDetectionPlugin(@unused reporter: viper.silver.reporter.Reporter,
+                           @unused logger: ch.qos.logback.classic.Logger,
+                           @unused config: viper.silver.frontend.SilFrontendConfig,
+                           fp: FastParser) extends SilverPlugin with ParserPluginTemplate {
+
+  import fp.{FP, ParserExtension, keyword}
+
+  /** Keyword used to define `unreachable` statement. */
+  private val unreachableKeyword: String = "unreachable"
+
+  /** Parser for `unreachable` statements. */
+  def unreachable[$: P]: P[PUnreachable] =
+    FP(keyword(unreachableKeyword)).map { case (pos, _) => PUnreachable()(pos) }
+
+  /** Add `unreachable` to the parser. */
+  override def beforeParse(input: String, isImported: Boolean): String = {
+    // Add new keyword
+    ParserExtension.addNewKeywords(Set[String](unreachableKeyword))
+    // Add new parser to the precondition
+    ParserExtension.addNewStmtAtEnd(unreachable(_))
+    input
+  }
+
+  override def beforeVerify(input: Program): Program = {
+    var refuteId = 0
+
+    // Add `refute false` at end of each method body
+    val transformedMethods = input.methods.map(method => {
+      val body = if (method.body.isEmpty) Seq() else Seq(method.body.get)
+      val bodyWithRefute = Option(Seqn(body :+ refuteFalse(method.pos, refuteId), Seq())(method.pos))
+
+      Method(method.name, method.formalArgs, method.formalReturns, method.pres,
+        method.posts, bodyWithRefute)(method.pos, method.info, method.errT)
+    })
+
+    // Add `refute false` after each inhale, at end of each loop body, at end of each if/else branch and before each goto
+    val methodsWithRefutes = transformedMethods.map(method => {
+      ViperStrategy.Slim({
+        case i@Inhale(expr) if i.info != Synthesized =>
```

Comment:
An AST node could potentially have multiple infos attached, so I think the check should be ``if i.info.getUniqueInfo[Synthesized.type].isEmpty``

Reply:
Thanks, I was not aware of this. -> b2f32b5d